azure-toolkit-libs/azure-toolkit-auth-lib/src/main/java/com/microsoft/azure/toolkit/lib/auth/Account.java

```diff
@@ -150,54 +164,115 @@ public String toString() {
         return StringUtils.join(details.toArray(), "\n");
     }
 
-    void authenticate() throws LoginFailureException {
-        try {
-            createTokenCredential();
-
-            initializeTenants();
+    protected Mono<Account> authenticate() {
+        if (!this.checkAvailable()) {
+            return Mono.error(new AzureToolkitAuthenticationException("Cannot find credential using auth method: " + this.getMethod()));
+        }
+        return Mono.fromCallable(this::createTokenCredential).flatMap(this::verifyAndGetTenantId).map(tenantAndToken -> {
+            if (tenantAndToken.getT2() instanceof MsalToken) {
+                if (StringUtils.isBlank(this.entity.getEmail())) {
+                    this.entity.setEmail(getEmailFromMsalToken(tenantAndToken.getT2()));
+                }
+                String refreshToken = getRefreshTokenFromMsalToken(tenantAndToken.getT2());
+                if (StringUtils.isNotEmpty(refreshToken)) {
+                    AzureEnvironment env = ObjectUtils.firstNonNull(this.entity.getEnvironment(), Azure.az(AzureCloud.class).getOrDefault());
+                    String authority = AzureEnvironmentUtils.getAuthority(env);
+                    //TODO: when vscode/vs login is added, make sure to change it here
+                    this.entity.setTenantCredential(new RefreshTokenTenantCredential(
+                            authority, IdentityConstants.DEVELOPER_SINGLE_SIGN_ON_ID, refreshToken
+                    ));
+                }
+            }
+            if (this.entity.getTenantIds() == null) {
+                this.entity.setTenantIds(tenantAndToken.getT1());
+            }
+            this.entity.setAuthenticated(true);
+            return this;
+        }).flatMap(ignore -> {
+            if (this.entity.getSubscriptions() == null) {
+                return initializeSubscriptions().then(Mono.just(this));
+            }
+            return Mono.just(this);
+        });
+    }
 
-            initializeSubscriptions();
+    private static String getEmailFromMsalToken(AccessToken accessToken) {
+        IAuthenticationResult result = ((MsalToken) accessToken).getAuthenticationResult();
 
-            entity.setAuthenticated(true);
-        } catch (AzureToolkitAuthenticationException e) {
-            entity.setLastError(e);
+        if (result != null && result.account() != null) {
+            return result.account().username();
         }
+        return null;
     }
 
-    List<String> listTenantIds(AzureEnvironment environment, TokenCredential credential) {
-        return AzureResourceManager.authenticate(credential
-                , new AzureProfile(environment)).tenants().list().stream().map(Tenant::tenantId).collect(Collectors.toList());
-    }
+    private String getRefreshTokenFromMsalToken(AccessToken accessToken) {
+        IAuthenticationResult result = ((MsalToken) accessToken).getAuthenticationResult();
+        if (result == null) {
+            return null;
```

Comment:
throw `AzureToolkitAuthenticationException` when result == null? As return null will make the account looks like authenticated but in fact not.

Reply:
will be replaced by an auth exception